metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala

```diff
@@ -1156,37 +1150,43 @@ class MetalsLspService(
     } else {
       publishSynthetics(path)
       worksheetProvider.onDidFocus(path)
-      buildTargets.inverseSources(path) match {
-        case Some(target) =>
-          val isAffectedByCurrentCompilation =
-            path.isWorksheet ||
-              buildTargets.isInverseDependency(
-                target,
-                compilations.currentlyCompiling.toList,
-              )
+      maybeCompileOnDidFocus(path).asJava
+    }
+  }
 
-          def isAffectedByLastCompilation: Boolean =
-            !compilations.wasPreviouslyCompiled(target) &&
-              buildTargets.isInverseDependency(
-                target,
-                compilations.previouslyCompiled.toList,
-              )
+  private def maybeCompileOnDidFocus(path: AbsolutePath) = {
+    buildTargets.inverseSources(path) match {
+      case Some(target) =>
+        val wasNeverCompiled = !compilations.wasEverCompiled(target)
+        def isAffectedByCurrentCompilation =
+          !compilations.isCurrentlyCompiling(target) && path.isWorksheet ||
+            buildTargets.isInverseDependency(
+              target,
+              compilations.currentlyCompiling.toList,
+            )
 
-          val needsCompile =
-            isAffectedByCurrentCompilation || isAffectedByLastCompilation
-          if (needsCompile) {
-            compilations
-              .compileFile(path)
-              .map(_ => DidFocusResult.Compiled)
-              .asJava
-          } else {
-            CompletableFuture.completedFuture(
-              DidFocusResult.AlreadyCompiled
+        def isAffectedByLastCompilation: Boolean =
+          !compilations.wasPreviouslyCompiled(target) &&
+            buildTargets.isInverseDependency(
+              target,
+              compilations.previouslyCompiled.toList,
```

Comment:
I wonder if instead we should have `onDidFocusBuildTarget` so recompile every time we switch build target. I worry that this would not work for example if for `A -> B` compile A, then some unrelated C and then open B which depends on A. Will we recompile B?

Reply:
No, we won't. I used the logic that was for used for `didFocus` but it is rather wanting. I think we can be smarter than what you're suggesting, since people might switch between the build targets a lot when only reading the files. We could keep timestamps of when the build target was compiled and trigger compilation only if any dependency was compiled since last time the build target in focus was compiled. But if you find this to be unnecessary elaborate and not worth it we can go with your suggestion instead.

Reply:
What you are suggesting would probably be better to do, but might also be more fragile. I would opt for the simpler solution to compilation when focused build target changes

Reply:
Okay, done.